src/types/permissions.ts

```diff
@@ -42,70 +54,138 @@ export enum PermissionType {
+    super();
+    this.amount = amount;
+  }
+
+  static from = (args: { amount: number }): GasPermission => new GasPermission(args);
+
+  serialize(serializer: Serializer): void {
+    serializer.serializeU16(this.amount);
+  }
+
+  static deserialize(deserializer: Deserializer): GasPermission {
+    const amount = deserializer.deserializeU16();
+    return GasPermission.from({ amount });
+  }
 }
 
-export interface NFTPermission {
-  type: PermissionType.NFT;
-  assetAddress: string;
-  capabilities: Record<NFTCapability, boolean>;
+export class NFTPermission extends Serializable {
+  readonly assetAddress: AccountAddress;
+
+  readonly capabilities: Record<NFTCapability, boolean>; // (string[], bool[])
+
+  constructor({
+    assetAddress,
+    capabilities,
+  }: {
+    assetAddress: AccountAddress;
+    capabilities: Record<NFTCapability, boolean>;
+  }) {
+    super();
+    this.assetAddress = assetAddress;
+    this.capabilities = capabilities;
+  }
+
+  static from = (args: { assetAddress: AccountAddress; capabilities: Record<NFTCapability, boolean> }): NFTPermission =>
+    new NFTPermission(args);
+
+  serialize(serializer: Serializer): void {
+    this.assetAddress.serialize(serializer);
+
+    const [capabilityKeys, capabilityValues] = Object.entries(this.capabilities).reduce(
+      ([keys, values], [key, value]) => [keys.concat(key), values.concat(value)],
+      [[] as string[], [] as boolean[]],
+    );
+    serializer.serializeStr(JSON.stringify(capabilityKeys));
+    serializer.serializeStr(JSON.stringify(capabilityValues));
+  }
+
+  static deserialize(deserializer: Deserializer): NFTPermission {
+    const assetAddress = AccountAddress.deserialize(deserializer);
+    const capabilityKeys = JSON.parse(deserializer.deserializeStr()) as string[];
+    const capabilityValues = JSON.parse(deserializer.deserializeStr()) as boolean[];
+    const capabilities = capabilityKeys.reduce(
+      (acc, key, i) => ({ ...acc, [key]: capabilityValues[i] }),
+      {} as Record<NFTCapability, boolean>,
+    );
+    return NFTPermission.from({ assetAddress, capabilities });
+  }
 }
 
-export interface NFTCollectionPermission {
-  type: PermissionType.NFTCollection;
-  collectionAddress: string;
+export class NFTCollectionPermission extends Serializable {
+  collectionAddress: AccountAddress;
+
   capabilities: Record<NFTCollectionCapability, boolean>;
-}
 
-export enum NFTCollectionCapability {
-  transfer = "transfer",
-  mutate = "mutate",
-}
+  constructor({
+    collectionAddress,
+    capabilities,
+  }: {
+    collectionAddress: AccountAddress;
+    capabilities: Record<NFTCollectionCapability, boolean>;
+  }) {
+    super();
+    this.collectionAddress = collectionAddress;
+    this.capabilities = capabilities;
+  }
 
-/**
- * Revoke permission types
- */
-export type RevokeFungibleAssetPermission = Pick<FungibleAssetPermission, "type" | "asset">;
-export type RevokeNFTPermission = Pick<NFTPermission, "type" | "assetAddress">;
```

Comment:
Note: I removed the revoking types here in favor of just using the base class. Might add these back in but modified after using the new style in connect.